### PR TITLE
Added modern dram mappings to memory controller.

### PR DESCRIPTION
**Pull request by maccoymerrell**
Modernized dram mappings have been added to the memory controller.
Because our model allows for arbitrary values for # of rows,banks, etc..., we cannot directly copy mappings from real machines.

This mapping is heavily based on the re'd zen4 mappings.

Each channel bit is xor'd with all row bits.

Bankgroup and bank bits are xor'd all at once, as follows.
[bank bits][bank group bits]
  ^ ^ ^ ^ ^ ^ ^ ^ ^ ^ ^ ^ ^ ^ ^
[row bits [bg bits + bk_bits:0]]
  ^ ^ ^ ^ ^ ^ ^ ^ ^ ^ ^ ^ ^ ^ ^
[row bits [(bg_bits + bk_bits)*(i+1)][(bg_bits + bk_bits)*i]]
 ....
 ....
 Until we are out of row bits to xor with.

New tests were added to test this, and ensure that no aliasing occurs in these mappings.

### CODE REVIEW

File: src/dram_controller.cc [hunk 1]
```diff
@@ -554,19 +554,55 @@ bool MEMORY_CONTROLLER::add_wq(const request_type& packet)
...
+  champsim::address row = champsim::address{get_row(address)};
+
+  //channel bits should be xor'd with each row bit
+  unsigned long c_bits = champsim::size(get<SLICER_CHANNEL_IDX>(address_slicer));
+  unsigned long r_bits = champsim::size(get<SLICER_ROW_IDX>(address_slicer));
+  for(unsigned long r = 0; r < r_bits; r++) {
+    for(unsigned long c = 0; c < c_bits; c++)
+      channel ^= row.slice_lower<champsim::data::bits{0}>().to<unsigned long>() << c;
+    row = champsim::address{row.to<unsigned long>() >> 1};
+  }
```
Comment (ngober) [inline src/dram_controller.cc:567]: Similar code shows up 3 times here. Is there a way to extract this into a `DRAM_ADDRESS_MAPPING::swizzle_bits(...)` function?

File: inc/dram_controller.h [hunk 2]
```diff
@@ -60,6 +60,7 @@ struct DRAM_ADDRESS_MAPPING {
   unsigned long get_bank(champsim::address address) const;
   unsigned long get_row(champsim::address address) const;
   unsigned long get_column(champsim::address address) const;
+  unsigned long swizzel_bits(champsim::address address, unsigned long segment_size, unsigned long segment_offset, unsigned long field, unsigned long field_bits) const;
```
Comment (ngober) [inline inc/dram_controller.h:63]:
```suggestion
  unsigned long swizzle_bits(champsim::address address, unsigned long segment_size, unsigned long segment_offset, unsigned long field, unsigned long field_bits) const;
```

File: src/dram_controller.cc [hunk 3]
```diff
@@ -552,21 +552,46 @@ bool MEMORY_CONTROLLER::add_wq(const request_type& packet)
...
 
+unsigned long DRAM_ADDRESS_MAPPING::swizzle_bits(champsim::address address, unsigned long segment_size, unsigned long segment_offset, unsigned long field, unsigned long field_bits) const {
+  champsim::address row = champsim::address{get_row(address)};
+  unsigned long permute_field = field;
+
+  unsigned long r_bits = champsim::size(get<SLICER_ROW_IDX>(address_slicer));
+  for(unsigned long r = 0; r < r_bits; r += segment_size) {
+    permute_field ^= row.slice(champsim::dynamic_extent{champsim::data::bits{segment_offset + field_bits},champsim::data::bits{segment_offset}}).to<unsigned long>();
+    row = champsim::address{row.to<unsigned long>() >> segment_size};
+  }
```
Comment (ngober) [inline src/dram_controller.cc:563]:
```suggestion
  for(champsim::dynamic_extent subextent{champsim::data::bits{segment_offset}, field_bits}; subextent.upper <= row.upper_extent(); subextent = champsim::dynamic_extent{subextent.upper, field_bits}) {
    permute_field ^= row.slice(subextent).to<unsigned long>();
  }
```
Again, you've taken a detour through the integral types when what you want to do can be achieved in the high-level interface. Also, I've here used an alternate constructor for `champsim::dynamic_extent` that takes a lower bit and a size.

File: src/dram_controller.cc [hunk 4]
```diff
@@ -552,21 +552,46 @@ bool MEMORY_CONTROLLER::add_wq(const request_type& packet)
   return false;
 }
 
+unsigned long DRAM_ADDRESS_MAPPING::swizzle_bits(champsim::address address, unsigned long segment_size, unsigned long segment_offset, unsigned long field, unsigned long field_bits) const {
```
Comment (ngober) [inline src/dram_controller.cc:555]:
```suggestion
unsigned long DRAM_ADDRESS_MAPPING::swizzle_bits(champsim::address address, unsigned long segment_size, champsim::data::bits segment_offset, unsigned long field, unsigned long field_bits) const {
```
This value is only ever used as this type, so you should make it part of the interface.

**Review by ngober — CHANGES_REQUESTED** [review]
Your bit swizzling takes too many descents into the domain of raw integer types. I've added some suggestions to avoid that.

File: src/dram_controller.cc [hunk 5]
```diff
@@ -552,21 +552,44 @@ bool MEMORY_CONTROLLER::add_wq(const request_type& packet)
   return false;
 }
 
+unsigned long DRAM_ADDRESS_MAPPING::swizzle_bits(champsim::address address, unsigned long segment_size, champsim::data::bits segment_offset, unsigned long field, unsigned long field_bits) const {
+  champsim::address_slice row{get<SLICER_ROW_IDX>(address_slicer), address};
+  unsigned long permute_field = field;
+
+  for(champsim::dynamic_extent subextent{champsim::data::bits{0}, segment_size}; subextent.upper <= row.upper_extent(); subextent = champsim::dynamic_extent{subextent.upper, segment_size}) {
+    permute_field ^= row.slice(subextent).slice(champsim::dynamic_extent{segment_offset,field_bits}).to<unsigned long>();
```
Comment (ngober) [inline src/dram_controller.cc:560]: I think the double slice is somewhat different than my suggestion. Could you explain how this works? The `address_slice::slice()` family of functions are relative to the domain of the slice. They are not absolute. So, a repeated slice is a subset of a subset of bits. Is that what you intended?

I see you also added some test cases to 751. Are these related?

Comment (maccoymerrell) [reply src/dram_controller.cc:560]: The segment being used from the row address and the bits being xor'd are not identical.

The function travels over the bits of the row address in segments of segment_size.
Bits are taken from [segment_offset + field_bits : segment_offset] of the segment to xor with the field.

The bits chosen for bankgroup and bank are a good example of why this is needed.
Say we have 2 bg bits and 2 bk bits. Zen4 takes the first two bits of the row and xor's it with the first 2 bg bits.
It then takes the next 2 bits from the row and xor's it with the bk bits. It then takes the next two bits and xor's it with the bg bits and so on, back and forth.
So swizzle_bits(address, bk_bits + bg_bits, 0, bankgroup, bg_bits) gives us the bankgroup index
while swizzle_bits(address, bk_bits + bg_bits, bg_bits, bank, bk_bits) gives us the bank index.

The testcases in 751 ensure that these mappings aren't broken. It tests to make sure that for a given row the channel, banks, and bankgroups never alias back into themselves and that each channel, bank, and bankgroup is visited exactly once as the row is traversed.

Comment (ngober) [reply src/dram_controller.cc:560]: OK, I understand. The segment size is the stride, and the segment offset is the location within that stride. That matches with what's given here, then.

File: src/dram_controller.cc [hunk 6]
```diff
@@ -552,21 +552,44 @@ bool MEMORY_CONTROLLER::add_wq(const request_type& packet)
   return false;
 }
 
+unsigned long DRAM_ADDRESS_MAPPING::swizzle_bits(champsim::address address, unsigned long segment_size, champsim::data::bits segment_offset, unsigned long field, unsigned long field_bits) const {
+  champsim::address_slice row{get<SLICER_ROW_IDX>(address_slicer), address};
+  unsigned long permute_field = field;
+
+  for(champsim::dynamic_extent subextent{champsim::data::bits{0}, segment_size}; subextent.upper <= row.upper_extent(); subextent = champsim::dynamic_extent{subextent.upper, segment_size}) {
```
Comment (ngober) [inline src/dram_controller.cc:559]:
```suggestion
  using namespace champsim::data::data_literals;
  for(champsim::dynamic_extent subextent{0_b, segment_size}; subextent.upper <= row.upper_extent(); subextent = champsim::dynamic_extent{subextent.upper, segment_size}) {
```

By the way, I don't know if you know that this syntax exists. If you want it, here it is.

**Review by ngober — APPROVED** [review]

**Comment by maccoymerrell** [comment]
I think we are good to go on this. I am happy with it.

**Comment by maccoymerrell** [comment]
Looks like something happened with curl.
